[FIX JENKINS-33770] Prevent unauthenticated user registration

kzantow opened the pull request:
Setup wizard didn't restrict access to all unnecessary paths prior to authentication.

This addresses: https://issues.jenkins-ci.org/browse/JENKINS-33770

@reviewbybees

kzantow committed:
JENKINS-33770 - not all paths restricted during SetupWizard

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -161,6 +163,11 @@ public HttpResponse doCompleteInstall() throws IOException, ServletException {
      * This filter will validate that the security token is provided
      */
     private final Filter FORCE_SETUP_WIZARD_FILTER = new Filter() {
+        private Pattern RESOURCE_PATTERN = Pattern.compile(".*[.](css|ttf|gif|woff|eot|png|js)");
```
Should maybe end in `$`?

kzantow replied:
@daniel-beck it's matching the full URL, try:
http://localhost:8080/adjuncts/asdf/asdf.js vs. http://localhost:8080/adjuncts/asdf/asdf.jsp

oleg-nenashev replied:
I suspect the method should be "static final"

kzantow replied:
There is no need, memory can be reclaimed up after use if it's not static

oleg-nenashev replied:
Anyway, not so important for the single-shot event

oleg-nenashev replied:
But non-static may cause some performance degradation due to the frequent compilation in tests [and real life?]

kzantow replied:
Not instantiated in tests frequently... rarely instantiated in normal Jenkins usage...

daniel-beck commented:
@jenkinsci/code-reviewers

jglick commented:
Hold on, I thought the whole point of using a real `SecurityRealm` during the setup wizard was so that we did not have to code-review fragile whitelists like this?

daniel-beck commented:
For some reason, the `admin` user created during startup is not considered a "real" user for purposes of `securityRealm/firstUser`, so it can be submitted successfully. If you try this in Jenkins 1.x after manually setting up one user and a security realm, the form submission will fail. I think investigating the criteria for that would be a better approach and should take care of @jglick's criticism.

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -54,6 +56,7 @@
  * @since 2.0
  */
+@Restricted(NoExternalUse.class) // use by Jelly
 public class SetupWizard {
```
I'm not convinced that SetupWizard can be used in Jelly only.
In the future plugin instances may hook on the wizard during the installation process in order to contribute some installation steps.

daniel-beck replied:
Probably more of a copy/paste issue. `DoNotUse.class` is the real Jelly use case.

jglick replied:
> In the future plugin instances may hook on the wizard during the installation process

Not sure how that would work. If and when we have such a requirement, an API can be designed accordingly.

oleg-nenashev commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -172,15 +179,24 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
+                    final String adjustedPath = req.getRequestURI().substring(req.getContextPath().length());
+                    final String destination;
+                    // if the user hasn't authenticated, don't allow anything other than the login submit and error page
+                    if (!isAdmin && !ALLOWED_PATHS.matcher(adjustedPath).matches()) {
```
This logic does not comply with the comment above. If we start providing new users during startup (e.g. useful for the DEV environment), then it will redirect non-admin users to the login page.

I think it makes sense to explicitly check the non-authenticated use-case

batmat commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -161,6 +163,11 @@ public HttpResponse doCompleteInstall() throws IOException, ServletException {
+        private Pattern ALLOWED_PATHS = Pattern.compile(
+            "\\Q/login\\E"
+            + "|\\Q/loginError\\E"
+            + "|\\Q/securityRealm/" + new HudsonPrivateSecurityRealm(true, false, null).getAuthenticationGatewayUrl() + "\\E");
```
+1 for static final (+here too, shoudn't it?)

kzantow replied:
It can be cleaned up after use if it's not static @batmat

batmat replied:
OK, I see, your goal is to lighten the weight of that class in memory IIUC, and because the class itself is likely to stay as JVM rarely unload classes?

kzantow replied:
@batmat right, the Jenkins instance has methods referring to the SetupWizard class, but after use, the SetupWizard instance is set to null and can be reclaimed, along with instance fields. Given the infrequent use of this class, I think having static fields here is unnecessary

jglick commented:
> For some reason, the `admin` user created during startup is not considered a "real" user for purposes of `securityRealm/firstUser`, so it can be submitted successfully. If you try this in Jenkins 1.x after manually setting up one user and a security realm, the form submission will fail.

Do not think I followed that. You are saying something about `HudsonPrivateSecurityRealm/firstUser.jelly`, or about something in `HudsonPrivateSecurityRealm.java`?

kzantow committed:
JENKINS-33770 - fix issue directly submitting firstUser page

kzantow commented:
@jglick it's a byproduct of me changing this around too much... I missed aligning this new user creation with the latest change... it's an issue in `HudsonPrivateSecurityRealm`, the `doCreateFirstAccount` method. I've addressed your concerns, hopefully and eliminated the URL pattern matching bit by addressing the actual issue of submitting the firstUser form during the setup wizard

kzantow commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -280,36 +287,14 @@ public void doCreateAccountByAdmin(StaplerRequest req, StaplerResponse rsp) thro
      * This can be run by anyone, but only to create the very first user account.
      */
     public void doCreateFirstAccount(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
-        boolean inSetup = !Jenkins.getInstance().getInstallState().isSetupComplete();
```
... this method was simply reverted to what it was before any of my 'security' changes

kzantow commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -261,16 +259,25 @@ private void loginAndTakeBack(StaplerRequest req, StaplerResponse rsp, User u) t
         checkPermission(Jenkins.ADMINISTER);
-        if(createAccount(req, rsp, false, "addUser.jelly")!=null) {
-            rsp.sendRedirect(".");  // send the user back to the listing page
+        User u = createAccount(req, rsp, false, addUserView);
```
... this is really not good, but given that the createAccount method handles changing views, I'm not sure if this could be made a lot better without more significant refactorings

daniel-beck replied:
I don't think this is an issue. Certainly not for 2.0.

jglick commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -131,11 +120,46 @@ public SetupWizard(Jenkins j) throws IOException, InterruptedException {
+                admin.delete(); // assume the new user may well be 'admin'
+            }
+            
+            User u = securityRealm.createAccountByAdmin(req, rsp, "/jenkins/install/SetupWizard/setupWizardFirstUser.jelly", j.getRootUrl());
```
Use `sendRedirect2` and pass `/`. Do not use `getRootUrl` internally. [Guide](https://wiki.jenkins-ci.org/display/JENKINS/Hyperlinks+in+HTML)

jglick commented:
Do not understand all of this, but :bee: so far as I can tell.

kzantow committed:
Relative URL for redirect

daniel-beck commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -131,11 +120,46 @@ public SetupWizard(Jenkins j) throws IOException, InterruptedException {
     /**
      * Gets the file used to store the initial admin password
      */
-    @Restricted(NoExternalUse.class) // use by Jelly
```
Restriction was removed for no apparent reason.

kzantow replied:
@daniel-beck - @Restricted(NoExternalUse) moved to class level

daniel-beck replied:
Great!

daniel-beck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -261,16 +259,25 @@ private void loginAndTakeBack(StaplerRequest req, StaplerResponse rsp, User u) t
+    /**
+     * Creates a user account. Requires {@link Jenkins#ADMINISTER}
+     */
+    public User createAccountByAdmin(StaplerRequest req, StaplerResponse rsp, String addUserView, String successView) throws IOException, ServletException {
```
`@Restricted`

kzantow replied:
No more/less restricted than `doCreateAccountByAdmin`, I think

jglick replied:
Both ought to be restricted I think, but certainly anything newly introduced should be.

daniel-beck replied:
@kzantow Methods named `doSomething` are implicit web methods, i.e. really exist only for Stapler. That's how we justify changing/deleting them as necessary even though we probably shouldn't from a backward compatibility POV.

Methods not fitting this pattern OTOH are always considered API.

kzantow replied:
> Methods named doSomething are implicit web methods, i.e. really exist only for Stapler. That's how we justify changing/deleting them

@daniel-beck don't like that implicit difference, really, but fair... I suppose it stems from the mix of UI with business logic, anyway hope this is all set now

daniel-beck replied:
@kzantow Well, ideally we add `@Restricted` to them when we change them anyway, but they existed long before those annotations.

daniel-beck commented:
:+1: from me.

@jenkinsci/code-reviewers Was changed significantly, PTAL.

jglick commented:
:bee:

daniel-beck commented:
One more `@Restricted` and this is done IMO.

kzantow committed:
Merge remote-tracking branch 'primary/2.0' into JENKINS-33770-security-token-not-always-required

kzantow committed:
@Restrict new method
